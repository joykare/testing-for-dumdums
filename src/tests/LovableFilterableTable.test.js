import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import fs from "fs";
import path from "path";
import serailizer from "jest-serializer-enzyme";
import LoveableFilterbleTable from "../LovableFilterableTable";
import { tableSchema } from "../App";

const SAMPLE_RESPONSE_FILE = path.join(__dirname, "../../sample-response.json");

const generateItems = () => {
  const response = fs.readFileSync(SAMPLE_RESPONSE_FILE);
  const json = JSON.parse(response);

  return json.slice(0, 30).map(item => ({
    ...item,
    isLoved: false
  }))
}

expect.addSnapshotSerializer(serailizer);

describe("LovableFilterableTable", () => {
  it("renders without crashing <smoke test>", () => {
    const items = [];
    const div = document.createElement("div");
    ReactDOM.render(
      <LoveableFilterbleTable
        items={items}
        schema={tableSchema}
      />,
    div);
  });

  let wrapper;

  describe("when given empty [items]", () => {
    const items = [];

    beforeEach(() => {
      wrapper = shallow(
        <LoveableFilterbleTable items={items} schema={tableSchema} />
      );
    });

    it("should still render search box", () => {
      expect(wrapper.find("input").exists()).toBe(true);
    });

    it("should have no table rows", () => {
      expect(wrapper.find("tbody > tr").exists()).toBe(false);
    });
  });

  describe("when given items", () => {
    const items = [
      { id: 1, name: "Bitcoin" },
      { id: 2, name: "Ethereum" },
      { id: 3, name: "Zcash" }
    ];

    beforeEach(() => {
      wrapper = shallow(
        <LoveableFilterbleTable items={items} schema={tableSchema} />
      );
    });

    it("should render tr elements", () => {
      expect(wrapper.find("tbody > tr").exists()).toBe(true);
    });

    it("should render the approriate number of tr elements", () => {
      expect(wrapper.find("tbody > tr").length).toBe(3);
    });

    it("should include the title of each item", () => {
      items.forEach(item => {
        expect(
          wrapper.containsMatchingElement(
            <td>
              {item.name}
            </td>
          )
        ).toBe(true);
      });
    });
  });

  describe("user entering search query", () => {
    let items;

    beforeEach(() => {
      items = generateItems();
      wrapper = shallow(
        <LoveableFilterbleTable items={items} schema={tableSchema} />
      );

      const searchBox = wrapper.find("input");
      searchBox.simulate("change", {target: {value: "coin"}});
    });

    it("render a subset of matching [items]", () => {
      const matching = items.filter(item => item.name.match(/coin/i));
      const notmatching = items.filter(item => !item.name.match(/coin/i));

      matching.forEach(match => {
        expect(
          wrapper.containsMatchingElement(
            <td>
              {match.name}
            </td>
          )
        ).toBe(true);
      });

      notmatching.forEach(match => {
        expect(
          wrapper.containsMatchingElement(
            <td>
              {match.name}
            </td>
          )
        ).toBe(false);
      });

      expect(wrapper.find("tbody").first().html()).toMatchSnapshot();
    });

    it("should filter items", () => {
      expect(wrapper.find("tbody > tr > .item-name").map(item => item.html())).toMatchSnapshot();
    });

    it("should filter items <with serializer>", () => {
      expect(wrapper.find("tbody")).toMatchSnapshot();
    });

    describe("users clears search query", () => {
      beforeEach(() => {
        const searchBox = wrapper.find("input");
        searchBox.simulate("change", {target: {value: ""}});
      });

      it("should render all [items] again", () => {
        expect(wrapper.find("tbody > tr").length).toEqual(items.length);
      });
    });
  });

  // Not too awesome better to test against user behaviour/interaction/expectations
  // describe("user entering search query <simulating vs setState()>", () => {
  //   let items;

  //   beforeEach(() => {
  //     items = generateItems();
  //     wrapper = shallow(<LoveableFilterbleTable items={items} schema={tableSchema} />);

  //     wrapper.instance().updateFilter("coin", items); //instance returns JS instance of the component so we can access the JS function directly
  //     wrapper.update();
  //   });

  //   it("should update state", () => {
  //     expect(wrapper.state("matches").map(match => match.name)).toMatchSnapshot();
  //   });
  // });
});