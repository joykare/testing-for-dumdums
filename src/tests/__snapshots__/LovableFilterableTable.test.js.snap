// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LovableFilterableTable user entering search query <simulating vs setState()> should update state 1`] = `
Array [
  "Bitcoin",
  "Ethereum",
  "Bitcoin Cash",
  "Ripple",
  "Litecoin",
  "NEM",
  "Dash",
  "Monero",
  "IOTA",
  "Ethereum Classic",
  "NEO",
  "BitConnect",
  "OmiseGO",
  "Qtum",
  "Lisk",
  "Stratis",
  "Waves",
  "Zcash",
  "Tether",
  "Hshare",
  "Bytecoin",
  "EOS",
  "Stellar Lumens",
  "BitShares",
  "TenX",
  "Steem",
  "MaidSafeCoin",
  "MCAP",
  "Augur",
  "Komodo",
]
`;

exports[`LovableFilterableTable user entering search query render a subset of matching [items] 1`] = `"<tbody><tr><td class=\\"item-name\\" style=\\"padding:5px;\\">Bitcoin</td><td class=\\"item-symbol\\" style=\\"padding:5px;\\">BTC</td><td class=\\"item-price_usd\\" style=\\"padding:5px;\\">4325.64</td><td class=\\"item-percent_change_1h\\" style=\\"padding:5px;\\">0.69</td><td class=\\"item-percent_change_24h\\" style=\\"padding:5px;\\">-3.93</td><td class=\\"item-percent_change_7d\\" style=\\"padding:5px;\\">0.15</td><td class=\\"item-heart\\"><i class=\\"fa fa-heart-o\\" aria-hidden=\\"true\\"></i></td></tr><tr><td class=\\"item-name\\" style=\\"padding:5px;\\">Bitcoin Cash</td><td class=\\"item-symbol\\" style=\\"padding:5px;\\">BCH</td><td class=\\"item-price_usd\\" style=\\"padding:5px;\\">539.182</td><td class=\\"item-percent_change_1h\\" style=\\"padding:5px;\\">0.78</td><td class=\\"item-percent_change_24h\\" style=\\"padding:5px;\\">-5.51</td><td class=\\"item-percent_change_7d\\" style=\\"padding:5px;\\">-10.68</td><td class=\\"item-heart\\"><i class=\\"fa fa-heart-o\\" aria-hidden=\\"true\\"></i></td></tr><tr><td class=\\"item-name\\" style=\\"padding:5px;\\">Litecoin</td><td class=\\"item-symbol\\" style=\\"padding:5px;\\">LTC</td><td class=\\"item-price_usd\\" style=\\"padding:5px;\\">67.1049</td><td class=\\"item-percent_change_1h\\" style=\\"padding:5px;\\">3.17</td><td class=\\"item-percent_change_24h\\" style=\\"padding:5px;\\">-11.17</td><td class=\\"item-percent_change_7d\\" style=\\"padding:5px;\\">7.34</td><td class=\\"item-heart\\"><i class=\\"fa fa-heart-o\\" aria-hidden=\\"true\\"></i></td></tr><tr><td class=\\"item-name\\" style=\\"padding:5px;\\">Bytecoin</td><td class=\\"item-symbol\\" style=\\"padding:5px;\\">BCN</td><td class=\\"item-price_usd\\" style=\\"padding:5px;\\">0.00190874</td><td class=\\"item-percent_change_1h\\" style=\\"padding:5px;\\">-0.75</td><td class=\\"item-percent_change_24h\\" style=\\"padding:5px;\\">-4.33</td><td class=\\"item-percent_change_7d\\" style=\\"padding:5px;\\">23.88</td><td class=\\"item-heart\\"><i class=\\"fa fa-heart-o\\" aria-hidden=\\"true\\"></i></td></tr><tr><td class=\\"item-name\\" style=\\"padding:5px;\\">MaidSafeCoin</td><td class=\\"item-symbol\\" style=\\"padding:5px;\\">MAID</td><td class=\\"item-price_usd\\" style=\\"padding:5px;\\">0.594994</td><td class=\\"item-percent_change_1h\\" style=\\"padding:5px;\\">2.37</td><td class=\\"item-percent_change_24h\\" style=\\"padding:5px;\\">-12.48</td><td class=\\"item-percent_change_7d\\" style=\\"padding:5px;\\">-2.96</td><td class=\\"item-heart\\"><i class=\\"fa fa-heart-o\\" aria-hidden=\\"true\\"></i></td></tr></tbody>"`;

exports[`LovableFilterableTable user entering search query should filter items <with serializer> 1`] = `
<tbody>
  <tr
    style={Object {}}
  >
    <td
      className="item-name"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      Bitcoin
    </td>
    <td
      className="item-symbol"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      BTC
    </td>
    <td
      className="item-price_usd"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      4325.64
    </td>
    <td
      className="item-percent_change_1h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      0.69
    </td>
    <td
      className="item-percent_change_24h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -3.93
    </td>
    <td
      className="item-percent_change_7d"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      0.15
    </td>
    <td
      className="item-heart"
    >
      <i
        aria-hidden="true"
        className="fa fa-heart-o"
        onClick={[Function]}
      />
    </td>
  </tr>
  <tr
    style={Object {}}
  >
    <td
      className="item-name"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      Bitcoin Cash
    </td>
    <td
      className="item-symbol"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      BCH
    </td>
    <td
      className="item-price_usd"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      539.182
    </td>
    <td
      className="item-percent_change_1h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      0.78
    </td>
    <td
      className="item-percent_change_24h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -5.51
    </td>
    <td
      className="item-percent_change_7d"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -10.68
    </td>
    <td
      className="item-heart"
    >
      <i
        aria-hidden="true"
        className="fa fa-heart-o"
        onClick={[Function]}
      />
    </td>
  </tr>
  <tr
    style={Object {}}
  >
    <td
      className="item-name"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      Litecoin
    </td>
    <td
      className="item-symbol"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      LTC
    </td>
    <td
      className="item-price_usd"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      67.1049
    </td>
    <td
      className="item-percent_change_1h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      3.17
    </td>
    <td
      className="item-percent_change_24h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -11.17
    </td>
    <td
      className="item-percent_change_7d"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      7.34
    </td>
    <td
      className="item-heart"
    >
      <i
        aria-hidden="true"
        className="fa fa-heart-o"
        onClick={[Function]}
      />
    </td>
  </tr>
  <tr
    style={Object {}}
  >
    <td
      className="item-name"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      Bytecoin
    </td>
    <td
      className="item-symbol"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      BCN
    </td>
    <td
      className="item-price_usd"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      0.00190874
    </td>
    <td
      className="item-percent_change_1h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -0.75
    </td>
    <td
      className="item-percent_change_24h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -4.33
    </td>
    <td
      className="item-percent_change_7d"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      23.88
    </td>
    <td
      className="item-heart"
    >
      <i
        aria-hidden="true"
        className="fa fa-heart-o"
        onClick={[Function]}
      />
    </td>
  </tr>
  <tr
    style={Object {}}
  >
    <td
      className="item-name"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      MaidSafeCoin
    </td>
    <td
      className="item-symbol"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      MAID
    </td>
    <td
      className="item-price_usd"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      0.594994
    </td>
    <td
      className="item-percent_change_1h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      2.37
    </td>
    <td
      className="item-percent_change_24h"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -12.48
    </td>
    <td
      className="item-percent_change_7d"
      style={
        Object {
          "padding": "5px",
        }
      }
    >
      -2.96
    </td>
    <td
      className="item-heart"
    >
      <i
        aria-hidden="true"
        className="fa fa-heart-o"
        onClick={[Function]}
      />
    </td>
  </tr>
</tbody>
`;

exports[`LovableFilterableTable user entering search query should filter items 1`] = `
Array [
  "<td class=\\"item-name\\" style=\\"padding:5px;\\">Bitcoin</td>",
  "<td class=\\"item-name\\" style=\\"padding:5px;\\">Bitcoin Cash</td>",
  "<td class=\\"item-name\\" style=\\"padding:5px;\\">Litecoin</td>",
  "<td class=\\"item-name\\" style=\\"padding:5px;\\">Bytecoin</td>",
  "<td class=\\"item-name\\" style=\\"padding:5px;\\">MaidSafeCoin</td>",
]
`;
